### PR TITLE
Build web, node, and production bundle

### LINKED ISSUE #7: Build web, node, and production bundle
<a href="https://github.com/m-cat"></img></a> **Issue by [m-cat](https://github.com/m-cat)**
_Tuesday Feb 23, 2021 at 19:56 GMT_
_Originally opened as https://github.com/NebulousLabs/skynet-js/pull/338_

----

I've moved to Webpack and created two separate targets for `node` and `dev`.

### Production build deploy

1. [x] The production web build should be deployed to our `skynet-js` HNS. I plan to include a `deploy.js` script in the repo which uses the Node version of the library itself to upload the production file and save the skylink to the registry. (#351)

### Update docs

1. [ ] Remove nodejs-skynet SDK from the docs (and archive repo)
2. [ ] Update nodejs examples
3. [ ] Add notes about differences in API between Node and Web (since the former lacks `File` and the latter lacks access to the local filesystem)

----

_**[m-cat](https://github.com/m-cat)** included the following code: https://github.com/NebulousLabs/skynet-js/pull/338/commits_

### PR DESCRIPTION
**Pull request by m-cat**
I've moved to Webpack and created two separate targets for `node` and `dev`.

Original PR: https://github.com/NebulousLabs/skynet-js/pull/338
Closes #7

### Production build deploy

1. [x] The production web build should be deployed to our `skynet-js` HNS. I plan to include a `deploy.js` script in the repo which uses the Node version of the library itself to upload the production file and save the skylink to the registry. (#351)

### Update docs

1. [ ] Remove nodejs-skynet SDK from the docs (and archive repo)
2. [ ] Update nodejs examples
3. [ ] Add notes about differences in API between Node and Web (since the former lacks `File` and the latter lacks access to the local filesystem)

### CODE REVIEW

File: scripts/deploy.js [hunk 1]
```diff
@@ -0,0 +1 @@
+const { SkynetClient } = require("");
```
Comment (kwypchlo) [inline scripts/deploy.js:1]: what is this used for?

File: src/client/index.ts [hunk 2]
```diff
@@ -109,6 +111,9 @@ export class SkynetClient {
    * @param [customOptions] Configuration for the client.
    */
   constructor(portalUrl: string = defaultPortalUrl(), customOptions: CustomClientOptions = {}) {
+    if (!portalUrl) {
+      portalUrl = defaultPortalUrl();
+    }
```
Comment (kwypchlo) [inline src/client/index.ts:116]: I don't think you need that, you already have that in the argument list
```suggestion
```

File: src/download/web.ts [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+export function openFile(this: SkynetClient, skylinkUrl: string, customOptions?: CustomDownloadOptions): string {
+  /* istanbul ignore next */
+  if (typeof skylinkUrl !== "string") {
+    throw new Error(`Expected parameter skylinkUrl to be type string, was type ${typeof skylinkUrl}`);
+  }
+
+  const opts = { ...defaultDownloadOptions, ...this.customOptions, ...customOptions };
+  const url = this.getSkylinkUrl(skylinkUrl, opts);
+
+  window.open(url, "_blank");
```
Comment (kwypchlo) [inline src/download/web.ts:82]: I think we might need to add noopener and noreferrer to this so it's safer

https://web.dev/external-anchors-use-rel-noopener/

https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/noreferrer
https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/noopener

File: src/download/web.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+  /* istanbul ignore next */
+  if (typeof domain !== "string") {
+    throw new Error(`Expected parameter domain to be type string, was type ${typeof domain}`);
+  }
+
+  const opts = { ...defaultDownloadHnsOptions, ...this.customOptions, ...customOptions };
+  const url = this.getHnsUrl(domain, opts);
+
+  // Open the url in a new tab.
+  window.open(url, "_blank");
```
Comment (kwypchlo) [inline src/download/web.ts:111]: same remark as with other `window.open`

File: src/upload/node.ts [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+    typeof response.data.bitfield !== "number"
+  ) {
+    throw new Error(
+      "Did not get a complete upload response despite a successful request. Please try again and report this issue to the devs if it persists."
+    );
+  }
+
+  const skylink = formatSkylink(response.data.skylink);
+  const merkleroot = response.data.merkleroot;
+  const bitfield = response.data.bitfield;
```
Comment (kwypchlo) [inline src/upload/node.ts:205]: optional tip: it's handy to use object destructuring assignment in such cases

```suggestion
  const { merkleroot, bitfield } = response.data;
```

**Review by kwypchlo — COMMENTED** [review]
did a quick scan, looks quite good 👍

File: scripts/deploy.js (hunk 1 above)
Comment (m-cat) [reply scripts/deploy.js:1]: I think this wasn't meant to be committed lol. The deploy script is a separate PR

**Pull request closed without merging**